File: py/2022/day20.py
from aoc import solve


def parse(s: str) -> list[int]:
    return [int(n) for n in s.splitlines()]


def decrypt_coordinates(data: list[int], key: int = 1, iterations: int = 1) -> int:
    values = [v * key for v in data]
    length = len(values)
    indexes = list(range(length))
    for _ in range(iterations):
        for i in range(length):
            pos = indexes.index(i)
            indexes.pop(pos)
            indexes.insert((pos + values[i]) % (length - 1), i)

    zero = indexes.index(values.index(0))
    return sum(values[indexes[(zero + n) % length]] for n in [1000, 2000, 3000])


parts = solve(
    lambda s: decrypt_coordinates(parse(s)),
    lambda s: decrypt_coordinates(parse(s), key=811589153, iterations=10),
    expect=(8721, 831878881825),
)

File: ts/2022/day20.ts
import { main } from 'lib/advent';
import { iter } from 'lib/iter';
import { allNumbers, range } from 'lib/util';

const KEY = 811589153;

function decryptCoordinates(xs: number[], key = 1, iterations = 1): number {
  const values = xs.map((x) => x * key);
  const indexes = range(0, values.length);
  for (let iter = 0; iter < iterations; ++iter) {
    for (let i = 0; i < values.length; ++i) {
      const pos = indexes.indexOf(i);
      indexes.splice(pos, 1);
      indexes.splice((pos + values[i]) % (values.length - 1), 0, i);
    }
  }

  const zero = indexes.indexOf(values.indexOf(0));
  return iter([1000, 2000, 3000])
    .map((n) => values[indexes[(zero + n) % values.length]])
    .sum();
}

main(
  (s) => decryptCoordinates(allNumbers(s)),
  (s) => decryptCoordinates(allNumbers(s), KEY, 10)
);
